contents/docs/hog/index.mdx

```diff
@@ -284,6 +286,47 @@ Hog's standard library includes the following functions and will expand. To see
 - `sha256Hex(arg: string): string`
 - `sha256HmacChainHex(arg: string[]): string`
 
+## Why Hog?
+
+Truth be told, we didn't set out to build our own programming language. We actually wanted to build the **HogVM**: a way to run HogQL expressions synchronously within our NodeJS ingestion pipeline. However, things escalated quickly.
+
+### HogVM
+
+The HogVM is a stack-based bytecode virtual machine (VM). It was built to power event filtering in our real-time ingestion flow.
+
+It's paired with a compiler that converts HogQL expressions like `properties.$screen_width < properties.$screen_height` into bytecode like `["_H", 1, 32, "$screen_height", 32, "properties", 1, 2, 32, "$screen_width", 32, "properties", 1, 2, 15, 35]`. The HogVM then executes that bytecode.
+
+The HogVM had a few design constraints:
+
+- Virtually no startup time (thousands of different expressions must run against millions of events every minute).
+- Fast enough to run hundreds (or thousands?) of these events per second on a single core.
```

Comment:
I'd leave numbers out here - "fast enough to filter the firehose of our event stream without drowning us in hosting costs" seems better (and is even a pun), if we're going to talk about speed.

Generally I'd shy away from _speed_ as a core design goal tbh - the realest wins (iiuc) with the hogvm are:
- startup time
- security/control of execution environment/resources/etc
- runtime flexibility (I'd add a bullet point here about being able to evaluate expressions in many contexts, and hogvm being embeddable in multiple languages/runtimes)
- HogQL eval (for the filtering usecase) - although this is more of a compiler feature than a HogVM one
- The async/serialisable stuff (I think you're right to break that into it's own section, below).

I don't think the hogvm is particularly highly optimised (open to being wrong here, but my understanding is most of our speed come from shelling out CPU intensive tasks to the underlying "native" runtime), nor does it need to be - the fact it's a win _in the face_ of the performance hit more notable, to me.

Reply:
All right, did some changes based on the above.

And indeed, you're right: HogVM is not particularly optimized, but it's fast enough. I'd bet there are some huge wins we can even have around the serialization code if we'd care to look.